hal/halx86/CMakeLists.txt

```diff
@@ -67,27 +67,32 @@ if(ARCH STREQUAL "i386")
     include(pcidata.cmake)
     include(legacy.cmake)
     include(up.cmake)
-    include(pic.cmake)
+    include(hal.cmake)
+    include(halapic.cmake)
+    include(halmps.cmake)
+    include(halacpi.cmake)
+    include(halaacpi.cmake)
+    include(halmacpi.cmake)
     include(xbox.cmake)
     include(pc98.cmake)
     add_subdirectory(minihal)
 
     remove_definitions(-DSARCH_XBOX)
     remove_definitions(-DSARCH_PC98)
 
-    # hal
-    add_hal(hal SOURCES up/halup.rc COMPONENTS generic legacy up pic)
-    add_hal(halacpi SOURCES acpi/halacpi.rc COMPONENTS generic acpi up pic)
-    add_hal(halapic SOURCES acpi/halacpi.rc COMPONENTS generic legacy up apic)
-    add_hal(halaacpi SOURCES acpi/halacpi.rc COMPONENTS generic acpi up apic)
+    # Legacy HALs
+    add_hal(hal SOURCES legacy/hal.rc COMPONENTS generic legacy hal)
+    add_hal(halapic SOURCES legacy/halapic.rc COMPONENTS generic legacy halapic)
+    add_hal(halmps SOURCES legacy/halmps.rc COMPONENTS generic legacy halmps)
+
+    # ACPI HALs
+    add_hal(halacpi SOURCES acpi/halacpi.rc COMPONENTS generic acpi halacpi)
+    add_hal(halaacpi SOURCES acpi/halaacpi.rc COMPONENTS generic acpi halaacpi)
+    add_hal(halmacpi SOURCES acpi/halmacpi.rc COMPONENTS generic acpi halmacpi)
+
```

Comment:
I'm not a fan of this way of cmake scripting, I appreciate the previous way, so it's something like...
```c
    # Legacy HALs
    add_hal(hal SOURCES legacy/hal.rc COMPONENTS generic legacy up pic)
    add_hal(halapic SOURCES legacy/halapic.rc COMPONENTS generic legacy up apic)
    add_hal(halmps SOURCES legacy/halmps.rc COMPONENTS generic legacy smp apic)

    # ACPI HALs
    add_hal(halacpi SOURCES acpi/halacpi.rc COMPONENTS generic acpi up pic)
    add_hal(halaacpi SOURCES acpi/halaacpi.rc COMPONENTS generic acpi up apic)
    add_hal(halmacpi SOURCES acpi/halmacpi.rc COMPONENTS generic acpi smp apic)
```
This way MPS definitions should be somewhere in `legacy` module. Note that it's not code suggestion, just a thought of how it should look like.

Reply:
I am not a fan :)
I just need the ability to individually compile a piece of code for a single HAL.
Example: halaacpi and halmacpi are almost the same except for spinlocks. This requires a separate compilation with the CONFIG_SMP key.
Another example: HaliHaltSystem() function - for ACPI HALs (halacpi, halaacpi, halmacpi) one option is used, for hal.dll another option, for halapic + halmps - this function is absent.
Another example: the arbiter is used only in hal.dll
It's like making LEGO from just 5 types of figures)
But I need a different type of figures...

Reply:
MPS definitions are not needed separately, as this code will also be the same for halapic and halmps. The only difference is that spinlock.c will compile with or without CONFIG_SMP.